pkgs/jni/lib/src/jni.dart

```diff
@@ -212,22 +213,77 @@ abstract final class Jni {
+  /// Flutter engine.
+  ///
+  /// Pass `PlatformDispatcher.instance.engineId` to the [engineId] field.
+  ///
+  /// The `ApplicationContext` is a long-lived singleton tied to the
+  /// application's lifecycle. It is safe to store and use from any thread
+  /// for non-UI tasks.
+  static JObject androidApplicationContext(int engineId) {
+    return JniPlugin.getApplicationContext(engineId);
   }
 
-  /// Returns current activity.
-  static JReference getCurrentActivity() =>
-      JGlobalReference(_bindings.GetCurrentActivity());
-
-  /// Get the initial classLoader of the application.
+  /// Retrieves the current Android `Activity` associated with a Flutter engine.
+  ///
+  /// The `engineId` can be obtained from `PlatformDispatcher.instance.engineId`
+  /// in Dart.
+  ///
+  /// **WARNING: This reference is volatile and must be used with care.**
+  ///
+  /// The Android `Activity` lifecycle is asynchronous. The `Activity` returned
+  /// by this function can become `null` or stale (destroyed) at any moment,
+  /// such as during screen rotation or when the app is backgrounded.
+  ///
+  /// To prevent native crashes, this function has two strict usage rules:
+  ///
+  /// 1. **Platform Thread Only**: It must *only* be called from the platform
+  ///     thread.
+  /// 2. **Synchronous Use Only**: The returned `JObject` must be used
+  ///     immediately and synchronously, with no asynchronous gaps (`await`).
   ///
-  /// This is especially useful on Android, where
-  /// JNI threads cannot access application classes using
-  /// the usual `JniEnv.FindClass` method.
-  static JReference getApplicationClassLoader() =>
-      JGlobalReference(_bindings.GetClassLoader());
+  /// Do not store the returned `JObject` in a field or local variable that
+  /// persists across an `await`.
+  ///
+  /// ---
+  ///
+  /// ### Correct Usage (Synchronous, "Get-and-Use"):
+  ///
+  /// ```dart
+  /// void safeCall() {
+  ///   // This is safe because the `Activity` is retrieved and used
+  ///   // in a single, unbroken, synchronous block.
+  ///   final activity = Jni.androidActivity(engineId);
+  ///   if (activity != null) {
+  ///     someGeneratedApi.doSomething(activity);
+  ///     activity.release();
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// ### **DANGEROUS** Usage (Asynchronous Gap):
+  ///
+  /// ```dart
+  /// Future<void> dangerousCall() async {
+  ///   // 1. Get the Activity (e.g., Activity "A")
+  ///   final activity = Jni.androidActivity(engineId);
+  ///
+  ///   // 2. An `await` occurs. The main thread is freed.
+  ///   //    While waiting, Android might destroy Activity "A" and create "B".
+  ///   await someOtherFuture();
+  ///
+  ///   // 3. CRASH: The code resumes, but `activity` is now a stale
+  ///   //    reference to the destroyed Activity "A".
+  ///   if (activity != null) {
+  ///     someGeneratedApi.doSomething(activity); // This will crash
+  ///     activity.release();
+  ///   }
+  /// }
+  /// ```
```

Comment:
These rules are similar to the rules for autorelease pools. I wonder if a similar API would work here, to make this less of a footgun:

```Dart
static void withAndroidActivity(int engineId, Function<void(JObject?)> callback) {
  final activity = JniPlugin.getActivity(engineId);
  callback(activity);
  activity?.release();
}
```

Looks a bit weird, and I'm not sure if the marginal safety improvement is worth it.